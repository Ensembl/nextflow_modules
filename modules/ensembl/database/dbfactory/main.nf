// See the NOTICE file distributed with this work for additional information
// regarding copyright ownership.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

process DATABASE_DBFACTORY {
    tag "${server.host}"
    label 'process_low'
    time '5min'

    conda "${moduleDir}/environment.yml"
    container "ensemblorg/ensembl-genomio:GenomioDockerRebuild_v1.5.0a"
    
    input:
        tuple val(server), val(filter_map)

    output:
        path "dbs.json", emit: dbs_meta_json
        path "versions.yml" , emit: versions

    when:
        task.ext.when == null || task.ext.when

    script:
        // Module specific vars:
        output_file = "dbs.json"
        dbname_re = filter_map.dbname_re ? "--db_regex ${filter_map}.dbname_re" : ''
        password_arg = server.password ? "--password ${server}.password" : ''

        // Prepare db_list to write in a file
        db_list = ""
        db_list_str = ""
        db_list_file = "db_list.tsv"

        if (filter_map.db_list) {
            db_list_str = filter_map.db_list.join("\n")
            db_list = "--db_list ${db_list_file}"
        }
        """
        echo "${db_list_str}" > ${db_list_file}

        database_factory --host '${server.host}' \
            --port '${server.port}' \
            --user '${server.user}' \
            ${password_arg} \
            --prefix '${filter_map.prefix}' \
            ${dbname_re} \
            ${db_list} \
            > ${output_file}
        
        echo -e -n "${task.process}:\n\tensembl-genomio: " > versions.yml
        python -c "import ensembl.io.genomio; print(ensembl.io.genomio.__version__)" >> versions.yml
        """

    stub:
        output_file = "dbs.json"
        """
        echo "{"species": "aaegL5", "division": "VectorBase", "release": 60}" > ${output_file}

        echo -e -n "${task.process}:\n\tensembl-genomio: " > versions.yml
        python -c "import ensembl.io.genomio; print(ensembl.io.genomio.__version__)" >> versions.yml
        """
}
