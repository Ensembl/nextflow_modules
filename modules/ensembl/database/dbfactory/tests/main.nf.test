// See the NOTICE file distributed with this work for additional information
// regarding copyright ownership.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// nf-core modules test database/dbfactory
nextflow_process {

    name "Test Process DATABASE_DBFACTORY"
    script "../main.nf"
    process "DATABASE_DBFACTORY"

    tag "modules"
    tag "modules_nfcore"
    tag "database"
    tag "database/dbfactory"

    test("Database dbfactory [stub]") {
            when {

                options "-stub"

                process {
                    // input[0] = Channel.of("mysql-host-test-prod")
                    //     .map { it -> [ host: it ] }
                    """
                    input[0] = Channel.of(
                        '{species: "aaegL5"}',
                        '{"division": "VectorBase"}',
                        '{"release": 60}'
                        )
                        .collectFile(name: 'dbs.json', newLine: true)
                        .map { server -> [ [ host: "mysql-host-test-prod" ], server ] }
                    """
                }
            }

            then {
                assert process.success
                assert process.trace.tasks().size() == 1

                assertAll(
                    { assert process.out.dbs_meta_json != null },
                    { assert snapshot(process.out).match() },
                    { assert process.out.dbs_meta_json.size() == 1 },
                    { assert process.out.dbs_meta_json.get(0).contains('dbs.json') }
                )
            }
        }
}