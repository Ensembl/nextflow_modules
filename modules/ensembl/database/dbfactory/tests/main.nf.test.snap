{
    "Database dbfactory [stub]": {
        "content": [
            {
                "0": [
                    "dbs.json:md5,476af329a36f4e5754d116a9f68b45f1"
                ],
                "1": [
                    "versions.yml:md5,4263356faf21f07aeb4f540175e50a18"
                ],
                "dbs_meta_json": [
                    "dbs.json:md5,476af329a36f4e5754d116a9f68b45f1"
                ],
                "versions": [
                    "versions.yml:md5,4263356faf21f07aeb4f540175e50a18"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-12-19T16:27:22.960883812"
    }
}